### PR TITLE
Even less warnings in tests

### PR DESCRIPTION
**Pull request by headtr1ck**
This PR removes several warnings from the tests and improves their typing on the way.

### CODE REVIEW

**Comment by headtr1ck** [comment]
Anyone know why on windows the async test do not capture the output?
I'm talking about the `test_async` and `test_serializable_locks` tests.

File: xarray/tests/test_dataarray.py [hunk 1]
```diff
@@ -4254,31 +4254,46 @@ def setup(self):
...
+        ),
+        pytest.param(
+            np.array([1.0, np.NaN, 2.0, np.NaN, -2.0, -4.0, 2.0]), 5, 2, 1, id="nan"
+        ),
+        pytest.param(
             np.array([1.0, np.NaN, 2.0, np.NaN, -2.0, -4.0, 2.0]).astype("object"),
             5,
             2,
             1,
+            marks=pytest.mark.filterwarnings(
```
Comment (mathause) [inline xarray/tests/test_dataarray.py:4271]: I don't mind this but I would probably just have slapped a decorator on the function, supressing the warning in general.

```python
@pytest.mark.filterwarnings("ignore:invalid value encountered in reduce:RuntimeWarning")
```

File: xarray/tests/test_dataarray.py [hunk 2]
```diff
@@ -5038,8 +5137,17 @@ def test_idxmin(self, x, minindex, maxindex, nanindex, use_dask) -> None:
             result7 = ar0.idxmin(dim="x", fill_value=-5j)
         assert_identical(result7, expected7)
 
-    @pytest.mark.parametrize("use_dask", [True, False])
-    def test_idxmax(self, x, minindex, maxindex, nanindex, use_dask) -> None:
+    @pytest.mark.parametrize(
+        "use_dask", [pytest.param(True, id="dask"), pytest.param(False, id="nodask")]
```
Comment (mathause) [inline xarray/tests/test_dataarray.py:5141]: Could probably do

```python
pytest.param("True", id="dask", marks=requires_dask)
```

and remove

```python
       if use_dask and not has_dask:
            pytest.skip("requires dask")
```

I think that's slightly more efficient than the current solution but let's leave that for another PR.

**Review by mathause — APPROVED** [review]
Some comments but happy to merge as is.

File: xarray/tests/test_dataarray.py [hunk 3]
```diff
@@ -5496,21 +5620,6 @@ def test_argmin_dim(
             },
             attrs=self.attrs,
         )
-        xindarr = np.tile(
```
[the remark this answers is not in the document]
Comment (headtr1ck) [reply xarray/tests/test_dataarray.py:5499]: I have to admit that mypy + flake8 found that xD

File: xarray/tests/test_variable.py [hunk 4]
```diff
@@ -2204,6 +2204,7 @@ def test_copy_index(self):
         super().test_copy_index()
 
     @pytest.mark.xfail
+    @pytest.mark.filterwarnings("ignore:elementwise comparison failed.*:FutureWarning")
```
[the remark this answers is not in the document]
Comment (headtr1ck) [reply xarray/tests/test_variable.py:2207]: Most warnings that I have encountered were correct and good to have. Like taking the mean of an all-nan array.
Very often also warnings about casting a NaN to int, which again I find useful because some users might not have that in mind.

In this specific case the future warning is also important to have as it will be a breaking change for most codes. Since it is used together with xfail, this test will fail anyway once the new behavior is implemented.

**Comment by mathause** [comment]
Thanks @headtr1ck!
